biosynseq/generate.py

```diff
@@ -0,0 +1,103 @@
+from gene_transformer.dataset import FASTADataset
+
+from torch.utils.data import DataLoader
+from Bio.Seq import Seq
+from pathlib import Path
+from argparse import ArgumentParser
+
+logger = logging.getLogger(__name__)
+
+
+def generate_fasta(cfg: ModelSettings, pt_path: str, fasta_path: str) -> dict:
+    """Given pt or deepspeed file, output generated sequences' fasta files."""
+    # obtain model
+    if Path(pt_path).suffix == ".pt":
+        # load pt file weights
+        model = DNATransformer.load_from_checkpoint(
+            checkpoint_path=pt_path, strict=False, cfg=cfg
+        )
+    else:
+        # load deepspeed weights
+        if cfg.load_from_checkpoint_dir is None:
+            raise ValueError("load_from_checkpoint_dir must be set in the config file.")
+        model = load_from_deepspeed(
+            cfg=cfg, checkpoint_dir=cfg.load_from_checkpoint_dir
+        )
+    model.cuda()
+    # generate non-redundant sequences
+    results = non_redundant_generation(model=model.model, tokenizer=model.tokenizer)
+    # turn unique sequences to fasta
+    unique_seqs = list(results.get("unique_seqs"))
+    seqs_to_fasta(seqs=unique_seqs, file_name=fasta_path)
+    return results
+
+
+def fasta_to_embeddings(
+    model_strategy: ModelLoadStrategy, fasta_path: str, embeddings_output_path: str
+) -> np.ndarray:
+    embeddings = inference(model_strategy, fasta_path, embeddings_output_path)
+    return embeddings
+
+
+if __name__ == "__main__":
+    parser = ArgumentParser(description="Generate sequences and/or embeddings.")
+    parser.add_argument("-c", "--config", type=str, required=True)
+    parser.add_argument(
+        "--mode",
+        default="get_fasta",
+        type=str,
+        required=True,
+        help="get_fasta or get_embeddings",
+    )
+    parser.add_argument("--pt_path", type=str, required=True)
+    parser.add_argument("--fasta_path", default="", type=str)
+    parser.add_argument(
+        "--embeddings_output_path", default="./embeddings.npy", type=Path
+    )
+    parser.add_argument(
+        "--embeddings_model_load", default="pt", type=str, help="deepspeed or pt"
+    )
+    args = parser.parse_args()
+    config = ModelSettings.from_yaml(args.config)
+
+    # set up torch environment
+    os.environ["TOKENIZERS_PARALLELISM"] = "true"
+    torch.set_num_threads(config.num_data_workers)  # type: ignore[attr-defined]
+    pl.seed_everything(0)
+
+    if args.mode == "get_fasta":
+        generate_fasta(cfg=config, pt_path=args.pt_path, fasta_path=args.fasta_path)
+    if args.mode == "get_embeddings":
+        if not args.fasta_path:
+            raise ValueError("Must provide a fasta file to run inference on.")
+
+        if args.embeddings_output_path.exists():
+            raise FileExistsError(
+                f"embeddings_output_path: {args.embeddings_output_path} already exists!"
+            )
+        if args.embeddings_model_load == "pt":
```

Comment:
Copy the model_strategy lines to run after the `pl.seed_everything(0)` line. Then pass the model_strategy object to your generate_fasta function, instead of passing the `pt_path`. Inside the generate_fasta function, simply run `model = model_strategy.get_model()` to get the model.